### PR TITLE
Run rclone as a mount helper program

### LINKED ISSUE #992: Feature Request: mount.rclone helper
It would be really handy if there was a mount.rclone wrapper that can be used in /etc/fstab. It would need to accept all the usual paremeters using slightly different syntax, e.g.

rclone --max-read-ahead 4M --read-only --config /path/to/config mount foo:/bar baz

would need to be expressed to the mount helper as:

mount.rclone -o max-read-ahead=4M,ro,config=/path/to/config foo:/bar baz

And, of course, mount.rclone would have to wait until the fs has completed mounting and then go into the background.

Perhaps this could be done by making rclone check whether the binary name it was invoked as is mount.rclone, and if so, alter it's parameter format expectations?

### PR DESCRIPTION
**Pull request by ivandeex**
#### What is the purpose of this change?

This PR will make rclonefs wrappers moot!

The PRedecessor #5593 solve(d) most problems that prevented rclone from running the mount in background i.e. working _like a generic unix mount_ program (`sshfs`, `mount.glusterfs` and others can do it out of the box), namely:
- have the main process prepare for mounting,
- spawn a background process to handle the mount,
- wait a little until the mount success or timeout (kill the background process if mount failed)
- don't block, exit with status code

The last incompatibility left is the way mount arguments are passed.
The core `/bin/mount` program will (in the basic scenario) get the `-t FSTYPE` argument then run the `/sbin/mount.FSTYPE` helper program passing it mount options. Automountd and systemd generally follow the same logic...
But rclone expects GNU-styled flags like `--key val` while mount options are passed as `-o key=val,...` (or `--opt=...`)!
This patch fills the last gap.

### Usage

Please symlink the rclone binary to `/sbin/mount.rclone` (and optionally `/usr/bin/rclonefs`) and **just**
- run something like
```
mount sftp1:subdir /mnt/data -t rclone -o vfs_cache_mode=writes,sftp_key_file=/path/to/pem
```
- or create systemd mount units (provide explicit `config=...,cache-dir=...` as mounts run without `HOME` !)
```
# /etc/systemd/system/mnt-data.mount
[Unit]
After=network-online.target
[Mount]
Type=rclone
What=sftp1:subdir
Where=/mnt/data
Options=rw,allow_other,args2env,vfs-cache-mode=writes,config=/etc/rclone.conf,cache-dir=/var/rclone
```
- optionally augmented by systemd automount unit
```
# /etc/systemd/system/mnt-data.automount
[Unit]
After=network-online.target
Before=remote-fs.target
[Automount]
Where=/mnt/data
TimeoutIdleSec=600
[Install]
WantedBy=multi-user.target
```
- or add in `/etc/fstab` a line like
```
sftp1:subdir /mnt/data rclone rw,noauto,nofail,_netdev,x-systemd.automount,args2env,vfs_cache_mode=writes,config=/etc/rclone.conf,cache_dir=/var/cache/rclone 0 0
```
- or use classic `Automountd` !

P.S. There was a ticket #2805 requesting the new `rclone mount --idle=X` flag but now we can do it in a standard way making it mostly moot (except on windows)...

### Implementation

~~This patch introduces new slice `fs.Args` which is by default a copy of `os.Args` but can be rewritten if rclone executable was invoked as `mount.rclone` or `rclonefs`.~~
**UPDATE**
This patch rewrites `os.Args` inplace in an early **init()** block if rclone executable was invoked as `mount.rclone` or `rclonefs`.

We use `os.Args[0]` rather than `os.Executable` to detect the name since the latter might resolve symlink while we want it unresolved.

Mount helpers usually lack the `HOME` or `RCLONE_CONFIG` environment variables, so config setup code in `fs/config` will require `fs.Args` to find a correct (probably converted) `--config FILE` running **very early** (most probably called from init blocks). Consequently the command line transformer also runs **super early** in the `init` block of the core `fs` package and avoids using other packages to **prevent import loops**. Utility functions `IsDaemon` and `IsMountHelper` were put directly in `fs` for this very reason.

Basically the conversion splits `-o` argument(s) by comma, replaces `_` by `-` and prepends `--` to transform `-o` options into Cobra flags. Options containing commas or spaces can be wrapped in single or double quotes (any quotes inside outer quotes should be doubled). The parser algorithm is based on the _connection string parser_ with a few command-line specific modifications.

The option syntax additionally includes a few pseudo mount options treated specially:
- `env.NAME=VALUE` will set an environment variable for this rclone instance and its children.
   This helps with Automountd and Systemd.mount which don't let set custom environment for mount helpers.
   Typically you will use `env.HTTPS_PROXY=proxy.host:3128` or `env.HOME=/root`
- `command=cmount` can be used to run any other command rather than default `mount` in the helper mode 
- `args2env` will pass mount options to the daemonized mount helper via `RCLONE_X` environment variables instead of CLI flags. This allows to hide secrets from `ps` or `pgrep`.
- `-o vv...` will be transformed into `--verbose=N`, just for convenience
- `x-systemd.automount`, `_netdev`, `nosuid` and alike are intended for automountd or systemd and ignored by rclone

#### Was the change discussed in an issue or in the forum before?

Based on #5593 (this PR includes preliminary patch of #5593 expected to annihilate at merge time)
Fixes #992

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by ivandeex** [comment]
@ncw
The latest `iva` build from https://github.com/ivandeex/rclone/releases already includes this patch so you can  play right now. I've run the patch on a staging box for a month or two before submitting.
Note the `-o args2env` mount flag that hides mount options from `pgrep -af rclone`.
I just need to write docs...

File: cmd/cmd.go [hunk 1]
```diff
@@ -273,7 +273,7 @@ func Run(Retry bool, showStats bool, cmd *cobra.Command, f func() error) {
 			break
 		}
 		if retryAfter := accounting.GlobalStats().RetryAfter(); !retryAfter.IsZero() {
-			d := retryAfter.Sub(time.Now())
+			d := time.Until(retryAfter)
```
Comment (ivandeex) [inline cmd/cmd.go:276]: Just to satisfy a warning in the VScode linter...

File: cmd/cmd.go [hunk 2]
```diff
@@ -458,7 +458,7 @@ func initConfig() {
 		})
 	}
 
-	if m, _ := regexp.MatchString("^(bits|bytes)$", *dataRateUnit); m == false {
+	if m, _ := regexp.MatchString("^(bits|bytes)$", *dataRateUnit); !m {
```
Comment (ivandeex) [inline cmd/cmd.go:461]: Just to satisfy a warning in the VScode linter...

File: fs/config/config.go [hunk 3]
```diff
@@ -246,9 +246,11 @@ func makeConfigPath() string {
 	// variable, since then we skip actually trying to create the default
 	// and report any errors related to it (we can't use pflag for this because
 	// it isn't initialised yet so we search the command line manually).
+	// Use processed command line from fs.Args so we can pick `-o config=...`
+	// if rclone was invoked as a mount helper.
 	_, configSupplied := os.LookupEnv("RCLONE_CONFIG")
 	if !configSupplied {
-		for _, item := range os.Args {
+		for _, item := range fs.Args {
 			if item == "--config" || strings.HasPrefix(item, "--config=") {
```
Comment (ivandeex) [inline fs/config/config.go:254]: This `fs/config` code can be used very early so we run conversion super early, in the `init` block of `fs` package.

**Review by ivandeex — COMMENTED** [review]
see notes in the code...

**Comment by ivandeex** [comment]
@ncw
The latest `iva` build from https://github.com/ivandeex/rclone/releases already includes this patch so you can  play right now. I've run the patch on a staging box for a month or two before submitting.
Note the `-o args2env` mount flag that hides mount options from `pgrep -af rclone`.

I don't yet call for review as I still need to put user notes from this PR's top post in the mount docs.

**Comment by ncw** [comment]
I had a quick skim of the code and it seems to do the right things :-)

I don't know if you saw my prototype for this? 78ffe3707792730354249524a8cfd8fb36bb8a57

Not sure there is anything you can learn from that, but I thought I'd put it out there.

It directly modifies `os.Args` rather than using an `fs.Args` which is less code to change.

**Comment by ivandeex** [comment]
> It directly modifies `os.Args` rather than using an `fs.Args` which is less code to change.

I wanted to keep a backup of original arguments just for debugging purposes.
If you think it's an overkill I can drop `fs.Args` and modify `os.Args` inplace.

**Comment by ivandeex** [comment]
> I don't know if you saw my prototype for this? [78ffe37](https://github.com/rclone/rclone/commit/78ffe3707792730354249524a8cfd8fb36bb8a57)

Not yet. Noted...

**Comment by ncw** [comment]
> > It directly modifies `os.Args` rather than using an `fs.Args` which is less code to change.
> 
> I wanted to keep a backup of original arguments just for debugging purposes.
> If you think it's an overkill I can drop `fs.Args` and modify `os.Args` inplace.

I'd say let's modify os.Args directly - we can always put a debug in with the old and the new values or something.

**Comment by ivandeex** [comment]
With my current load I cannot do much before 1.57

**Comment by ivandeex** [comment]
> let's modify os.Args directly - we can always put a debug in with the old and the new values or something.

done

**Review by ncw — COMMENTED** [review]
I think this code looks great - thank you.

However it could do with some docs - you wrote some nice ones in the PR - maybe cut and paste into `cmd/mountlib/help.go`? (Though I think that file is getting too long and we should probably break it out into its own docs page say `docs/content/mount.md` at some point...)

File: fs/config.go [hunk 4]
```diff
@@ -179,6 +180,11 @@ func NewConfig() *ConfigInfo {
 		if arg == "--log-level=DEBUG" || (arg == "--log-level" && len(os.Args) > argIndex+1 && os.Args[argIndex+1] == "DEBUG") {
 			c.LogLevel = LogLevelDebug
 		}
+		if strings.HasPrefix(arg, "--verbose=") {
```
Comment (ivandeex) [inline fs/config.go:183]: @ncw 
Thank you very much for review.

Do you mind if I put mountlib documentation in a separate commit?

Also I'd like to note that this early `--verbose=x` support in fs/config.go isn't directly connected with mount helper cli translator? I could split it in yet another commit. Do you mind?

Thank you

**Review by ivandeex — COMMENTED** [review]
two small questions...

File: fs/config.go (hunk 4 above)
Comment (ivandeex) [reply fs/config.go:183]: @ncw 
Thank you very much for review.

Do you mind if I put mountlib documentation in a separate commit?

Also I'd like to note that this early `--verbose=x` parser in `fs/config.go` isn't directly connected with mount helper cli translator and may be split it in yet another commit. Do you agree? Or better keep it squashed?

Thank you

**Comment by ivandeex** [comment]
@ncw @albertony 

Please be aware that I documented rclone mount helper in a dedicated commit https://github.com/rclone/rclone/pull/5594/commits/bd82c19dc0084094b86d2919d4f58283601e2c62 partially including the `make commanddocs` output for your reference.

Also updated the related wiki article https://github.com/rclone/rclone/wiki/rclone-mount-helper-script

**Comment by ivandeex** [comment]
@ncw @albertony 

Please be aware that I documented rclone mount helper in a dedicated commit https://github.com/rclone/rclone/pull/5594/commits/bd82c19dc0084094b86d2919d4f58283601e2c62 (https://github.com/rclone/rclone/commit/a065fb23e5238f5252002799905d455265e94998 on master) partially including the `make commanddocs` output for your reference.

Also updated the related wiki article https://github.com/rclone/rclone/wiki/rclone-mount-helper-script

**Comment by ncw** [comment]
Thank looks great, thank you.

**Comment by ivandeex** [comment]
Cool.
I'm moving on with submitting FTP stuff accumulated this summer.
